js/main.js

```diff
@@ -1,24 +1,48 @@
+const sheetsUrl =  'https://docs.google.com/spreadsheets/d/e/2PACX-1vTTUD7zK1YRnkzoaVoD33VGMUywCaFhS9gRZXa2_SAEb5PnIne7YWxWqy8k1vYvdc5ocT75_KleDmSX/pub?gid=0&single=true&output=csv';
+const cacheKey = 'v2-' + sheetsUrl;
+
+/// Synchronous function to parse raw data from google sheets to a form we are happy with
+/// Expects data in format produced by d3's CSV loading functions
+function parseSheetCSV(data) {
+    return data.map(d => {
+        // Phone and type should be lists, since there can be multiples separated by /
+        d['phone'] = d['phone'].split('/').map(n => n.trim());
+        d['type'] = d['type'].split('/').map(n => n.trim());
+        return d;
+    });
+}
+
+/// Return initial data for the contact list
+/// We return localStorage cached data if found, and count that as 'loaded'
+/// If not, we return empty list and mark it as loading
+function initialData() {
+    const cachedData = localStorage.getItem(cacheKey);
+    if (cachedData) {
+        return {
+            contactList: parseSheetCSV(JSON.parse(cachedData)),
+            isLoading: false
+        }
+    }
+    return {
+        contactList: [],
+        isLoading: true
+    };
+}
 
 // document.addEventListener("DOMContentLoaded", function() {
 const list = new Vue({
     "el": "#contact-list",
-    data:{
-        contactList: [...dataList],
-        isLoading: true
-    },
+    data: initialData(),
     created()
     {
-       
+
     },
     mounted(){
-        d3.csv('https://docs.google.com/spreadsheets/d/e/2PACX-1vTTUD7zK1YRnkzoaVoD33VGMUywCaFhS9gRZXa2_SAEb5PnIne7YWxWqy8k1vYvdc5ocT75_KleDmSX/pub?gid=0&single=true&output=csv').then(function (data) {
-            dataList = data.map(d => {
-                d['phone'] = d['phone'].split('/').map(n => n.trim());
-                d['type'] = d['type'].split('/').map(n => n.trim());
-                return d;
-            });
-            list.contactList = [...dataList];
+        d3.csv(sheetsUrl).then(function (data) {
```

Comment:
Just to be clear, this still loads the data from the Google Sheet even if it finds it in `localStorage` and then updates the `localStorage` and the UI when it does, right?

If so could you add a comment to this effect here.

Reply:
@prtksxna done!